### ISSUE TITLE
Create a generic way of hardcoding the test runner

### ISSUE THREAD

**Issue by janko-m** [janko-m/vim-test#185]
In some languages there are many test runners which use the same file pattern for test files (notably JavaScript and Python), and it's sometimes hard for test.vim to determine which test runner it should run. To work around that, for some of these test runners we are checking whether `test#{language}#runner` variables is declared and if it matches, to allow users to manually set which test runner they want to use.

In order to avoid the same `test#{language}#runner` logic in each of the test runners, and to allow users to always hardcode the test runner, we want to extract this logic out of the individual test runners and make it generic.

**Comment by codeinabox** [comment]
I guess one way to get around this is check for runners in the project dev dependencies.

**Comment by janko-m** [comment]
@codeinabox Yes, runner implementations should still attempt to do everything they can to determine the correct test runner, which includes checking dev dependencies. This would just be a way to override that and set the test runner manually, e.g. if the runner is installed only globally, or if test.vim has any other flaw in its logic.

**Comment by codeinabox** [comment]
@janko-m based on your earlier points, on what level will a user set the runner - globally, per buffer or both?

> However, for the sake of consistency we can leave it as it is, and remove it once we implement #185. To clarify, after #185 I plan that we stop activating a JavaScript runner if only the pattern matches and the global executable exists (without it existing in package.json). Considering tha then users will be able to selecting runners manually (instead of having to rely on test.vim to always correctly detect it), we can require that pattern matches and the runner is present in package.json, or the runner is hardcoded via a variable.

**Comment by janko-m** [comment]
@codeinabox Globally, I was thinking that the user can set the `g:test#runner` variable to the runner they want to use, in which case any pattern matching and other checks are skipped. We could then add a method which checks for both `g:test#runner` and the current `g:test#{language}#runner` variables, and in both cases returns the value (they would be treated the same).

**Comment by codeinabox** [comment]
Sounds good. I could try picking this up in a couple of weeks when I am back from holidays.

**Comment by codeinabox** [comment]
@janko-m I might be able to pick this up for JavaScript at the very least. Would this variable override the test file pattern or just the runner detection? I am thinking something like this 

```viml
function! test#javascript#mocha#test_file(file) abort
  return a:file =~# g:test#javascript#mocha#file_pattern
	  \ && (g:test#javascript#runner == 'mocha' || test#javascript#has_package('mocha') )
endfunction
```

**Comment by janko-m** [comment]
@codeinabox Just the runner detection. The idea is that then individual runners could remove the `test#{language}#runner` logic, because it would be detected and handled globally by test.vim for all runners.

**Comment by codeinabox** [comment]
👌 

In regards to me testing this feature with specs I'd have to create a fixture without package.json. Is doing that for just one JavaScript runner say Mocha sufficient?

**Comment by janko-m** [comment]
@codeinabox Actually, you're right, the runner *should* probably be able to override the file pattern. So that when someone's pattern doesn't match, that instead of having to override the pattern they can just hardcode the runner. So, in this case the runner is hardcoded the whole `test#{language}#{runner}#test_file(file)` function would be skipped (i.e. considered to return true).

Yeah, you could test without `package.json`, but I think it would be enough to test that a runner is used even if the file you're running the tests for doesn't match the file pattern of that runner.

**Comment by codeinabox** [comment]
I'd be inclined just to make it override the runner detection. Judging by all the issues the problem is not the file pattern but the runner detection, which has improved with package.json based detection, but this would give more granular control.

If people need to override the file pattern then this could be implemented later.

**Comment by janko-m** [comment]
@codeinabox Yes, I think we mean the same thing, the `test#{language}#{runner}#test_file()` function *is* the runner detection.

Overriding the pattern is already possible, because each runner have the file pattern string as a variable which can be overriden. I just wanted it to be clear that the new `g:test#runner` variable would override even that, more specifically it wouldn't care about the whole result of `test#{language}#{runner}#test_file()` function, which of course makes sense.

**Comment by janko-m** [comment]
@codeinabox Extracting the filename pattern into a variable is something that all built-in test.vim runners have, but it's not mandatory for any custom strategy that a user might have locally, so I would prefer not to assume existence of that variable.

Notice that the code snippet you posted wouldn't have the behaviour that we just discussed, because we said the `variable_override_check` would *override* file pattern matching; if a test runner selected in `test#runner` or `test#{language}#runner`, that test runner should be activated regardless of whether the test runner matches.

However, now I see that this actually isn't what we want, and that's why I was delaying the implementation, and I think your idea is a good one. However, do you think you can make `test#base#test_file()` assume existence of `test#{language}#{runner}#file_pattern` variable *only* if `test#runner`/`test#{language}#runner` is set? That way we eliminate any chance of breaking things for anyone.

**Comment by codeinabox** [comment]
I'll try to create a PR when I get a chance. If it passes all the existing specs is that a sufficient check for not breaking things for anyone?

**Comment by janko-m** [comment]
@codeinabox All built-in runners already have the `test#{language}#{runner}#file_pattern` variable, so it won't be tested in an automated way, but I don't think it's necessary to make an automated test for that.

**Comment by verkholantsev** [comment]
Is there any workaround to set specific runner to JavaScript project?

**Comment by codeinabox** [comment]
Not yet, as I didn't get to finish building this feature. Till then it's
determined by the runner it finds in your package.json
